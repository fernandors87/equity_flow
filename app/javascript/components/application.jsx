import '../src/application.scss'
import DatePicker from '../components/datepicker'
import MonthlyExpenses from 'components/monthly_expenses'
import React from 'react'
import { bindActionCreators } from 'redux'
import { connect } from 'react-redux'
import { fetchAccounts } from 'actions/accounts-actions'
import { fetchSplits } from 'actions/splits-actions'
import { updateChrono } from 'actions/chrono-actions'

export class Application extends React.Component {

  constructor(props) {
    super(props)

    this._onDateChange = this._onDateChange.bind(this)
  }

  componentDidMount() {
    const { start, end } = this.props.chrono
    this.props.fetchAccounts()
    this.props.fetchSplits(start, end)
  }

  _onDateChange(startDate, endDate) {
    const chronoAction = this.props.updateChrono(startDate, endDate)
    const { start, end } = chronoAction.payload.chrono
    this.props.fetchSplits(start, end)
  }

  render () {
    const { accounts, chrono, splits } = this.props
    const { start, end } = chrono
    return (
      <div>
        <DatePicker startDate={ start } endDate={ end } onDateChange={ this._onDateChange } />
        <MonthlyExpenses startDate={ start } endDate={ end } accounts={ accounts } splits={ splits } type="expense" aggregation="month" />
      </div>
    )
  }
}

const mapStateToProps = function(state) {
  return {
    accounts: state.accounts,
    chrono: state.chrono,
    splits: state.splits,
  }
}

const mapDispatchToProps = function(dispatch) {
  return bindActionCreators({
    fetchAccounts,
    fetchSplits,
    updateChrono,
  }, dispatch)
}

export default connect(mapStateToProps, mapDispatchToProps)(Application)
