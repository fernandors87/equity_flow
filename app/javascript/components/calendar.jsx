import '../css/calendar.scss'
import { Button, Glyphicon } from 'react-bootstrap'
import PropTypes from 'prop-types'
import { Range } from 'immutable'
import React from 'react'
import moment from 'moment'

export default class Calendar extends React.Component {

  constructor(props) {
    super(props)
    this.state = {
      date: props.date,
      month: moment(props.date).startOf('month')
    }

    this._onDateClick = this._onDateClick.bind(this)
    this._onMonthChange = this._onMonthChange.bind(this)
  }

  componentWillReceiveProps(nextProps) {
    const { date } = nextProps
    const month = moment(date).startOf('month')
    this.setState({ date, month })
  }

  _onDateClick(date) {
    const { onDateClick } = this.props
    if (onDateClick) {
      onDateClick(date)
    } else {
      const month = moment(date).startOf('month')
      this.setState({ date, month })
    }
  }

  _onMonthChange(offset) {
    this.setState(prevState => ({ month: moment(prevState.month).add(offset, 'month') }))
  }

  render() {
    const { date, month } = this.state
    const { className } = this.props

    const shortMonth = month.format('YYYY-MM')

    const weekNames = moment.weekdaysMin()
      .map(w => w[0].toUpperCase())
      .map((w, i) => <td key={ i }>{w}</td>)

    const startOfCalendar = moment(month).startOf('week')
    const calendarDates = Range(0, 42).map(i => moment(startOfCalendar).add(i, 'days'))
    const calendarDatesByWeek = calendarDates.groupBy(d => d.week())
    const calendarCellsByWeek = calendarDatesByWeek.map(dates => dates.map((calendarDate) => {
        const key = calendarDate.toString()
        const CellCss = 'day'
        const onClick = () => this._onDateClick(calendarDate)
        const buttonCss = calendarDate.isSame(date, 'day') ? 'selected' : 'unselected'
        const buttonStyle = buttonCss === 'selected' ? 'primary' : 'default'
        const button = <Button className={ buttonCss } bsStyle={ buttonStyle }>{calendarDate.date()}</Button>
        return <td key={ key } className={ CellCss } onClick={ onClick }>{button}</td>
      })).valueSeq()

    const calendarRows = calendarCellsByWeek.map((cells, i) => <tr key={ i }>{cells}</tr>)

    return (
      <table className={ `calendar ${className}` }>
        <thead>
          <tr>
            <td>
              <Button className="prev-month" onClick={ () => this._onMonthChange(-1) }>
                <Glyphicon glyph="triangle-left" />
              </Button>
            </td>
            <td colSpan="5">{shortMonth}</td>
            <td>
              <Button className="next-month" onClick={ () => this._onMonthChange(1) }>
                <Glyphicon glyph="triangle-right" />
              </Button>
            </td>
          </tr>
          <tr>
            {weekNames}
          </tr>
        </thead>
        <tbody>
          {calendarRows}
        </tbody>
      </table>
    )
  }
}

Calendar.propTypes = {
  date: PropTypes.instanceOf(moment).isRequired,
  onDateClick: PropTypes.func,
  className: PropTypes.string
}
