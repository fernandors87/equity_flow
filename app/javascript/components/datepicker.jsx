import '../css/datepicker.scss'
import { FormControl, FormGroup, Glyphicon, OverlayTrigger, Popover } from 'react-bootstrap'
import Calendar from './calendar'
import PropTypes from 'prop-types'
import React from 'react'
import moment from 'moment'

function DatePickerPopover({ startDate, endDate, onDateChange }) {
  const onDateChangeFn = (variable) => (date) => onDateChange({ [variable]: date })
  return (
    <Popover className="datepicker-popover" id="datepicker-popover">
      <Calendar className="calendar-start" date={ startDate } onDateClick={ onDateChangeFn('startDate') }/>
      <Calendar className="calendar-end" date={ endDate } onDateClick={ onDateChangeFn('endDate') }/>
    </Popover>
  )
}

function DatePickerInput({ container, startDate, endDate }) {
  const dateFormat = 'LL'
  const value = `${startDate.format(dateFormat)} - ${endDate.format(dateFormat)}`
  return <FormControl className="datepicker-input" type="text" readOnly container={ container } value={ value } />
}

export default class DatePicker extends React.Component {

  constructor(props) {
    super(props)
    this.state = {
      startDate: props.startDate,
      endDate: props.endDate
    }
    this._onDateChange = this._onDateChange.bind(this)
    this._onPopoverExit = this._onPopoverExit.bind(this)
  }

  componentWillReceiveProps(nextProps) {
    this.setState({
      startDate: nextProps.startDate,
      endDate: nextProps.endDate
    })
  }

  _onDateChange({ startDate, endDate }) {
    this.setState(prevState => {
      const prevStart = prevState.startDate
      const prevEnd = prevState.endDate
      const startChanged = startDate && startDate.isSameOrBefore(prevEnd, 'day')
      const endChanged = endDate && endDate.isSameOrAfter(prevStart, 'day')
      return {
        startDate: startChanged ? startDate : prevStart,
        endDate: endChanged ? endDate : prevEnd
      }
    })
  }

  _onPopoverExit() {
    const { startDate, endDate } = this.state
    const callback = this.props.onDateChange || (() => {})
    callback(startDate, endDate)
  }

  render() {
    const { startDate, endDate } = this.state
    const popover = DatePickerPopover({ startDate, endDate, onDateChange: this._onDateChange })

    return (
      <FormGroup className="datepicker">
        <OverlayTrigger
          rootClose
          className="datepicker-overlay"
          trigger="click"
          placement="bottom"
          overlay={ popover }
          onExiting={ this._onPopoverExit }>
          <div>
            <DatePickerInput container={ this } startDate={ startDate } endDate={ endDate }/>
            <FormControl.Feedback>
              <Glyphicon glyph="calendar" />
            </FormControl.Feedback>
          </div>
        </OverlayTrigger>
      </FormGroup>
    )
  }
}

DatePicker.propTypes = {
  startDate: PropTypes.instanceOf(moment).isRequired,
  endDate: PropTypes.instanceOf(moment).isRequired,
  onDateChange: PropTypes.func
}
