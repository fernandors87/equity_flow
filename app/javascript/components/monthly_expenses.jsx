import { List } from 'immutable'
import React from 'react'
import moment from 'moment'
import { range } from 'support/date_utils'
import { sum } from 'support/stats_utils'

export function Header({ months }) {
  const cells = months.map(m => <th key={ m.format() }>{ m.format('YYYY-MM') }</th>)
  return (
    <thead>
      <tr>
        <th>Account</th>
        { cells }
        <th>Sum</th>
        <th>Avg</th>
      </tr>
    </thead>
  )
}

export function Body({ accounts, months = List(), splits = List() }) {
  const rows = accounts.sortBy(a => a.name).map(a => {
    const splitsOfAccount = splits.filter(s => s.account_id === a.id)
    return <Row key={ a.id } label={ a.name } months={ months } splits={ splitsOfAccount } />
  })
  return (
    <tbody>
      { rows }
    </tbody>
  )
}

export function Footer({ months, splits }) {
  return (
    <tfoot>
      <Row label="Total" months={ months } splits={ splits } />
    </tfoot>
  )
}

export function Row({ label, months, splits }) {
  const splitsByMonth = splits.groupBy(s => moment(s.date).startOf('month'))
  const total = sum(splits.map(s => s.value)).toFixed(0)
  const avg = (total / months.size).toFixed(0)
  const cells = months.map(m => {
    const values = splitsByMonth.get(m, List()).map(s => s.value)
    const formatted = sum(values).toFixed(0)
    return <td className="value" key={ m.format() }>{ formatted }</td>
  })

  return (
    <tr>
      <td className="category">{ label }</td>
      { cells }
      <td className="summary total">{ total }</td>
      <td className="summary avg">{ avg }</td>
    </tr>
  )
}

export default class MonthlyExpenses extends React.Component {

  constructor(props) {
    super(props)
  }

  render() {
    const { accounts, splits, startDate, endDate } = this.props
    const months = range(startDate, endDate, 'month')
    const expenseAccounts = accounts.filter(a => a.type === 'expense')
    const expenseIds = expenseAccounts.map(a => a.id)
    const splitsOfRange = splits.filter(s => s.date.isBetween(startDate, endDate, 'day', '[]'))
    const expenseSplits = splitsOfRange.filter(s => expenseIds.contains(s.account_id))
    return (
      <table className="monthly-expenses table table-hover">
        <Header months={ months } />
        <Body accounts={ expenseAccounts } months={ months } splits={ expenseSplits } />
        <Footer months={ months } splits={ expenseSplits } />
      </table>
    )
  }
}
