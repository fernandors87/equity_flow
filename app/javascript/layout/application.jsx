import 'css/application'
import ApplicationContent from 'layout/application-content'
import ApplicationNavbar from 'layout/application-navbar'
import React from 'react'

export default class Application extends React.Component {

  render() {
    return (
      <div className="application">
        <ApplicationNavbar />
        <ApplicationContent view="home" />
      </div>
    )
  }
}
