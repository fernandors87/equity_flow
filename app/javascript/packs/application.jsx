import React from 'react'
import ReactDOM from 'react-dom'

import moment from 'moment'
import {Set} from 'immutable'

import * as API from './../src/api'
import '../src/application.scss'
import TransactionSummary from './transaction_summary'

class Application extends React.Component {

  constructor(props) {
    super(props)
    this.state = {
      accounts: Set(),
      splits: Set()
    }
  }

  componentDidMount() {
    const accounts = API.accounts.list()
    const splits = API.splits.list()
    Promise.all([accounts, splits]).then(([accounts, splits]) => {
      this.setState({accounts, splits})
    })
  }

  render () {
    return (
      <div>
        <TransactionSummary
          accounts={this.state.accounts}
          splits={this.state.splits}
          startDate={moment("2017-07-01").utc()}
          endDate={moment("2017-12-31").utc()}
          />
      </div>
    )
  }
}

window.api = API

document.addEventListener('DOMContentLoaded', () => {
  ReactDOM.render(
    <Application />,
    document.body.appendChild(document.createElement('div')),
  )
})
