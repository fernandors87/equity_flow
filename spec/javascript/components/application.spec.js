import { Application } from 'components/application'
import Chrono from 'models/chrono'
import { List } from 'immutable'

const chrono = new Chrono('2017-09-11T00:00:00.000+00:00', '2018-01-07T00:00:00.000+00:00')
const emptyAction = () => {}

it('should render the component', () => {
  const wrapper = mount(
    <Application
      accounts={ List() }
      chrono={ chrono }
      splits={ List() }
      fetchAccounts={ emptyAction }
      fetchSplits={ emptyAction }
      updateChrono={ emptyAction } />)

  expect(wrapper.html()).toMatchSnapshot()
})

it('should request remote accounts', () => {
  const accountsAction = sinon.stub()
  const splitsAction = sinon.stub()
  mount(
    <Application
      accounts={ List() }
      chrono={ chrono }
      splits={ List() }
      fetchAccounts={ accountsAction }
      fetchSplits={ splitsAction }
      updateChrono={ emptyAction } />)

  expect(accountsAction.callCount).toEqual(1)
  expect(splitsAction.callCount).toEqual(1)
})

describe('_onDateChange', () => {

  const chronoAction = sinon.stub().returns({ payload: { chrono } })

  it('should request remote splits', () => {
    const splitsAction = sinon.stub()
    const wrapper = mount(
      <Application
        accounts={ List() }
        chrono={ chrono }
        splits={ List() }
        fetchAccounts={ emptyAction }
        fetchSplits={ splitsAction }
        updateChrono={ chronoAction } />)

    wrapper.instance()._onDateChange(chrono.start, chrono.end)
    expect(splitsAction.callCount).toEqual(2)
  })

  it('should update the chrono', () => {
    const wrapper = mount(
      <Application
        accounts={ List() }
        chrono={ chrono }
        splits={ List() }
        fetchAccounts={ emptyAction }
        fetchSplits={ emptyAction }
        updateChrono={ chronoAction } />)

    wrapper.instance()._onDateChange(chrono.start, chrono.end)
    expect(chronoAction.callCount).toEqual(2)
  })
})
