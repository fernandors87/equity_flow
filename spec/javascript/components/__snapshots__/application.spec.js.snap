// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render the component 1`] = `"<div><div class=\\"datepicker form-group has-feedback\\"><div><input readonly=\\"\\" container=\\"[object Object]\\" type=\\"text\\" class=\\"datepicker-input form-control\\" value=\\"September 11, 2017 - January 7, 2018\\"><span class=\\"form-control-feedback glyphicon glyphicon-calendar\\"></span></div></div><table class=\\"monthly-expenses table table-hover\\"><thead><tr><th>Account</th><th>2017-09</th><th>2017-10</th><th>2017-11</th><th>2017-12</th><th>2018-01</th><th>Sum</th><th>Avg</th></tr></thead><tbody></tbody><tfoot><tr><td class=\\"category\\">Total</td><td class=\\"value\\">0</td><td class=\\"value\\">0</td><td class=\\"value\\">0</td><td class=\\"value\\">0</td><td class=\\"value\\">0</td><td class=\\"summary total\\">0</td><td class=\\"summary avg\\">0</td></tr></tfoot></table></div>"`;
